src/NUnitFramework/framework/Internal/MessagePumpStrategy.cs

```diff
@@ -0,0 +1,121 @@
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Threading;
+
+#if NET20 || NET35 || NET40 || NET45
+using System.Windows.Forms;
+#if !NET20
+using System.Windows.Threading;
+#endif
+#endif
+
+namespace NUnit.Framework.Internal
+{
+    internal abstract class MessagePumpStrategy
+    {
+        public abstract void WaitForCompletion(AwaitAdapter awaitable);
+
+        public static MessagePumpStrategy FromCurrentSynchronizationContext()
+        {
+#if NET20 || NET35 || NET40 || NET45
+            var context = SynchronizationContext.Current;
+
+            if (context is WindowsFormsSynchronizationContext)
+                return WindowsFormsMessagePumpStrategy.Instance;
+#if !NET20
+            if (context is DispatcherSynchronizationContext)
+                return WpfMessagePumpStrategy.Instance;
+#endif
+#endif
+
+            return NoMessagePumpStrategy.Instance;
+        }
+
+        private sealed class NoMessagePumpStrategy : MessagePumpStrategy
+        {
+            public static readonly NoMessagePumpStrategy Instance = new NoMessagePumpStrategy();
+            private NoMessagePumpStrategy() { }
+
+            public override void WaitForCompletion(AwaitAdapter awaitable)
+            {
+                awaitable.BlockUntilCompleted();
+            }
+        }
+
+#if NET20 || NET35 || NET40 || NET45
+        private sealed class WindowsFormsMessagePumpStrategy : MessagePumpStrategy
+        {
+            public static readonly WindowsFormsMessagePumpStrategy Instance = new WindowsFormsMessagePumpStrategy();
+            private WindowsFormsMessagePumpStrategy() { }
+
+            public override void WaitForCompletion(AwaitAdapter awaitable)
+            {
+                var context = SynchronizationContext.Current;
+
+                if (!(context is WindowsFormsSynchronizationContext))
+                    throw new InvalidOperationException("This strategy must only be used from a WindowsFormsSynchronizationContext.");
+
+                if (awaitable.IsCompleted) return;
+
+                // Post rather than running now so that if there has been a race condition and it completed
+                // after the IsCompleted check, it will wait until the application runs *before* shutting it down.
+                // Otherwise Application.Exit is a no-op and we would then proceed to do Application.Run and never return.
+                context.Post(
+                    state => ((AwaitAdapter)state).OnCompleted(Application.Exit),
+                    state: awaitable);
+
+                Application.Run();
+            }
+        }
+
+#if !NET20
+        private sealed class WpfMessagePumpStrategy : MessagePumpStrategy
```

Comment:
:bulb: `DispatcherMessagePumpStrategy`

Reply:
You're right, but Dispatcher is such a generic name. Is it used for anything but WPF?

Reply:
Worse than generic... it has a specific use already in the codebase. The WorkItemDispatcher is the thing that dispatches the work items that constitute the implementation of a test. We have three different dispatchers that do this in different ways.

Reply:
This is a private nested class, so if you're seeing the name you're also seeing that it's coupled to DispatcherSynchronizationContext. I'm less worried there and more just intent on... um, bikeshedding. 😎